src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -313,11 +313,33 @@ internal static bool TryGetMinimalCSharpRange(RazorCodeDocument codeDocument, Ra
         CancellationToken cancellationToken)
     {
         var parameter = new ProvideSemanticTokensRangesParams(textDocumentIdentifier, documentVersion, csharpRanges, correlationId);
+        ProvideSemanticTokensResponse? csharpResponse;
+        if (_languageServerFeatureOptions.UsePreciseSemanticTokenRanges)
+        {
+            csharpResponse = await GetCsharpResponseAsync(parameter, CustomMessageNames.RazorProvidePreciseRangeSemanticTokensEndpoint, cancellationToken).ConfigureAwait(false);
 
-        var csharpResponse = await _languageServer.SendRequestAsync<ProvideSemanticTokensRangesParams, ProvideSemanticTokensResponse>(
-            CustomMessageNames.RazorProvideSemanticTokensRangeEndpoint,
-            parameter,
-            cancellationToken).ConfigureAwait(false);
+            // Likely the server doesn't support the new endpoint, fallback to the original one
+            if (csharpResponse?.Tokens is null && csharpRanges.Length > 1)
+            {
+                var minimalRange = new Range
+                {
+                    Start = csharpRanges.First().Start,
+                    End = csharpRanges.Last().End
```

Comment:
```suggestion
                    Start = csharpRanges[0].Start,
                    End = csharpRanges[^1].End
```

I'm not actually sure if this makes a difference on arrays. @DustinCampbell probably knows

Reply:
Just my 2 cents: Functionally no, and the performance benefit is small. `First()` and `Last()` will ultimately end up going through `IList<T>`. However, I would argue that it's better for code readability. Whenever I see `First()` or `Last()`, I always have to check to see if it's going to end up going through `IEnumerable<T>` to be certain that it's going to be fast. Using the indexers in this way is not only (slightly) cheaper, but also clearer when read.